### PR TITLE
Implement min size constraint for splitter

### PR DESCRIPTION
**Pull request by pyroxymat**
I am not sure about the API still for the split constraints. But this is a draft of how to implement them. You can check it out in the "split_demo" example.

Closes issue #540

### CODE REVIEW

**Comment by pyroxymat** [comment]
Okay. I simplified the API additionally, I change the splitter demo to only have a single splitter now. Since it doesn't make a lot of sense to have a nested splitter in case the min-sizes are really off (the min_size for the right 2 is together the same as the minsize for the left split).

I guess a good solution at some point might be to move to a splitter where you can have multiple splits? This is how QT does it as far as I know.

**Comment by Zarenor** [comment]
I think we should index the panes left-to-right and top-to-bottom. So to min-size them independently, we have `first_min`(`min_first`?) and `min_second`. I could imagine max-sizes too, but I'm not sure how we handle the possibility of being asked to split a container too large for our maximum sizes if they're both/all set.

I had gone for 'simple, fast, and composable' with the single-splitter-only design. It handles, I think, the 80% case and has much simpler code than a multiway split. I also think filling out the constraints (mins and maxes) makes having to split several times more tolerable. I think being able to set the split point in units (pixels?) gets us the rest of the way to workable.

However, I wouldn't be opposed to creating a separate multisplitter, or altering this one with a vec of parameters, and so an arbitrary number of splits. Whichever we think is more appropriate. I think in that case agreeing to indexing in a direction would matter.

File: druid/src/widget/split.rs [hunk 1]
```diff
@@ -65,6 +67,14 @@ impl<T> Split<T> {
         self.split_point = split_point;
         self
     }
+    /// Set the minimum size for both sides of the split
+    /// The value must be atleast 0.0
```
Comment (cmyr) [inline druid/src/widget/split.rs]:
```suggestion
    /// The value must be greater than or equal to `0.0`.
```

File: druid/src/widget/split.rs [hunk 2]
```diff
@@ -92,33 +102,33 @@ impl<T> Split<T> {
             }
         }
     }
+    fn calculate_limits(&self, size: Size) -> (f64, f64) {
+        let size_in_split_direction = match self.split_direction {
```
Comment (cmyr) [inline druid/src/widget/split.rs]: Just fyi I believe you can do, `self.split_direction.major(size)` to get the relevant 'main axis length' for this direction.

File: druid/src/widget/split.rs [hunk 3]
```diff
@@ -316,3 +355,15 @@ impl<T: Data> Widget<T> for Split<T> {
         self.child2.paint_with_offset(paint_ctx, &data, env);
     }
 }
+
+// Move to std lib clamp as soon as https://github.com/rust-lang/rust/issues/44095 lands
+fn clamp(mut x: f64, min: f64, max: f64) -> f64 {
+    assert!(min <= max);
+    if x < min {
```
Comment (cmyr) [inline druid/src/widget/split.rs]: Not a blocker, but I would probably just write this as `x.max(min).min(max)`?

Comment (pyroxymat) [reply druid/src/widget/split.rs]: I just did it how Rust (nightly) did it, to keep compatibility in edge-cases. There is quite some [discussion](https://github.com/rust-lang/rfcs/pull/1961#issuecomment-289893089) going on what is the proper way to write clamp. Seems like at edge cases weird things happens with these small (seemingly) similar implementations.

File: druid/src/widget/split.rs (hunk 2 above)
Comment (pyroxymat) [reply druid/src/widget/split.rs]: Nice that is cleaner! FYI: I had to make the function public (for the crate) though.

File: druid/src/widget/split.rs (hunk 1 above)
Comment (pyroxymat) [reply druid/src/widget/split.rs]: Seems a bit clearer what happens in case of 0.0 indeed :)

File: druid/src/widget/split.rs (hunk 2 above)
Comment (pyroxymat) [reply druid/src/widget/split.rs]: I just noticed I actually had them the wrong way around as well...

Comment (pyroxymat) [reply druid/src/widget/split.rs]: For anyone wondering what the answer is, as seen in a [previous discussion](https://xi.zulipchat.com/#narrow/stream/147926-druid/topic/split.20vertical.2Fhorizontal.20confusion):

The axis tells you the axis of the splitter itself.

File: druid/src/widget/split.rs [hunk 4]
```diff
@@ -65,6 +67,14 @@ impl<T> Split<T> {
         self.split_point = split_point;
         self
     }
+    /// Set the minimum size for both sides of the split
```
Comment (cmyr) [inline druid/src/widget/split.rs]:
```suggestion
    /// Builder-style method to set the minimum size for both sides of the split.
    ///
```

**Pull request merged**